flink-test-utils-parent/flink-connector-testing-framework/src/main/java/org/apache/flink/connectors/test/common/testsuites/TestSuiteBase.java

```diff
@@ -0,0 +1,430 @@
+    @Tag("failover")
+    @DisplayName("Test TaskManager failure")
+    public void testTaskManagerFailure(TestEnvironment testEnv, ExternalContext<T> externalContext)
+            throws Exception {
+
+        checkEnvironmentIsControllable(testEnv);
+
+        final Collection<T> testRecordsBeforeFailure = externalContext.generateTestData();
+        final SourceSplitDataWriter<T> sourceSplitDataWriter = externalContext.createSourceSplit();
+        sourceSplitDataWriter.writeRecords(testRecordsBeforeFailure);
+
+        final StreamExecutionEnvironment env = testEnv.createExecutionEnvironment();
+
+        env.enableCheckpointing(50);
+        final DataStreamSource<T> dataStreamSource =
+                env.fromSource(
+                                externalContext.createSource(Boundedness.CONTINUOUS_UNBOUNDED),
+                                WatermarkStrategy.noWatermarks(),
+                                "Tested Source")
+                        .setParallelism(1);
+
+        // Since DataStream API doesn't expose job client for executeAndCollect(), we have
+        // to reuse these part of code to get both job client and result iterator :-(
+        // ------------------------------------ START ---------------------------------------------
+        TypeSerializer<T> serializer = dataStreamSource.getType().createSerializer(env.getConfig());
+        String accumulatorName = "dataStreamCollect_" + UUID.randomUUID();
+        CollectSinkOperatorFactory<T> factory =
+                new CollectSinkOperatorFactory<>(serializer, accumulatorName);
+        CollectSinkOperator<T> operator = (CollectSinkOperator<T>) factory.getOperator();
+        CollectResultIterator<T> iterator =
+                new CollectResultIterator<>(
+                        operator.getOperatorIdFuture(),
+                        serializer,
+                        accumulatorName,
+                        env.getCheckpointConfig());
+        CollectStreamSink<T> sink = new CollectStreamSink<>(dataStreamSource, factory);
+        sink.name("Data stream collect sink");
+        env.addOperator(sink.getTransformation());
+        final JobClient jobClient = env.executeAsync("TaskManager Failover Test");
+        iterator.setJobClient(jobClient);
+        // -------------------------------------- END ---------------------------------------------
+
+        checkSingleSplitRecords(
+                testRecordsBeforeFailure.iterator(), iterator, testRecordsBeforeFailure.size());
+
+        // -------------------------------- Trigger failover ---------------------------------------
+        final ClusterControllable controller = (ClusterControllable) testEnv;
+        controller.triggerTaskManagerFailover(jobClient, () -> {});
+
+        JobStatusUtils.waitForJobStatus(
+                jobClient, Collections.singletonList(JobStatus.RUNNING), Duration.ofSeconds(30));
+
+        final Collection<T> testRecordsAfterFailure = externalContext.generateTestData();
```

Comment:
this could be the exact same data as before right? couldn't the test then pass by just rereading the first data points?

Reply:
Depends on implementation of ```generateTestData```. For example in Kafka implementations it'll generate different random strings in different method calls. I'd like to call this method again to generate a new dataset after triggering failure.

Reply:
I think `generateTestData` should take a seed as a parameter so that we are sure that the implementation does not generate the same data. The test may be a false positive otherwise.